File: Range Queries/Distinct_Values_Queries.py
import sys,math,random
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations

# functions #
MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(2**62)
def gcd(a,b):
    if a%b==0:
        return b
    else:
        return gcd(b,a%b)
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
##

#String hashing : sh, fenwick sortedlist : fsortl, Number : numtheory, SparseTable : SparseTable
#bucket sorted list : bsortl, segment tree(lazy propogation) : SegmentTree, bootstrap : bootstrap
#binary indexed tree : BIT, segment tree(point updates) : SegmentPoint, Convex Hull : hull
#Combinatorics : pnc, Diophantine Equations : dpheq, Graphs : graphs, DSU : DSU
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

class BIT:
    #Faster than segment tree so use if possbile
    def __init__(self, x):
        """transform list into BIT"""
        self.bit = x
        for i in range(len(x)):
            j = i | (i + 1)
            if j < len(x):
                x[j] += x[i]
    def update(self, idx, x):
        """updates bit[idx] += x"""
        #basically after that number greater greater than x will be added
        while idx < len(self.bit):
            self.bit[idx] += x
            idx |= idx + 1
    def query(self, end):
        """calc sum(bit[:end])"""
        #gives sum of element before end
        x = 0
        while end:
            x += self.bit[end - 1]
            end &= end - 1
        return x
    def findkth(self, k):
        """Find largest idx such that sum(bit[:idx]) <= k"""
        idx = -1
        for d in reversed(range(len(self.bit).bit_length())):
            right_idx = idx + (1 << d)
            if right_idx < len(self.bit) and k >= self.bit[right_idx]:
                idx = right_idx
                k -= self.bit[idx]
        return idx + 1
    
def solve():
    n,q = list(map(int, sys.stdin.readline().split()))
    L = list(map(int, sys.stdin.readline().split()))
    d = defaultdict(list)
    
    for i in range(n-1,-1,-1):
        d[w(L[i])].append(i)
    k = [0 for i in range(n)]
    for i in d:
        k[d[i][-1]] = 1
        d[i].pop()
    seg = BIT(k)
    L2 = []
    for i in range(q):
        L1 = list(map(int, sys.stdin.readline().split()))
        L2.append((L1[0]-1,L1[1]-1,i))
    L2.sort()
    fans = [0 for i in range(q)]
    k3 = 0
    # def func(seg):
    #     for i in range(n):
    #         print(seg.query(i,i+1),end=' ')
    #     print()
    for st,ed,pos in L2:
        for l in range(k3,st):
            if d[w(L[l])]:
                seg.update(d[w(L[l])].pop(),1)
        k3 = st
        fans[pos] = (seg.query(ed+1)-seg.query(st))
    print(*fans,sep='\n')
    #st = sys.stdin.readline().strip()
solve()

File: Range Queries/Distinct_Values_Queries.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
using namespace std;
#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define print(arr) for (auto it : arr){cout<<it<<" ";}cout<<endl;
#define len(arr) arr.size()
#define printf(x) cout << x << endl;
#define printm(map) cout<<"{";for(auto it: map){cout<<it.first<<":"<<it.second<<",";};cout<<"}"<<endl;

// const int MOD = 998244353;
const int MOD = 1e9 + 7;

int gcd(int a, int b) {
    if (b == 0) return a;
    return gcd(b, a % b);
}

int lcm(int a, int b) {
    return a / gcd(a, b) * b;
}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable
// Segment Tree(lazy propogation): SegmentTree
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie/Treap: Tries
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
// Persistent Segment Tree: perseg, FreqGraphs: bgraph
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

class BIT {
public:
    int n;
    vector<int> bit;

    BIT(vector<int>& arr) {
        n = arr.size();
        bit = arr;
        for (int i = 0; i < n; ++i) {
            int j = i | (i + 1);
            if (j < n) bit[j] += bit[i];
        }
    }

    void update(int idx, int x) {
        while (idx < n) {
            bit[idx] += x;
            idx |= idx + 1;
        }
    }

    int query(int end) {
        int x = 0;
        while (end > 0) {
            x += bit[end - 1];
            end &= end - 1;
        }
        return x;
    }

    int findkth(int k) {
        int idx = -1;
        for (int d = 63 - __builtin_clzll(n); d >= 0; --d) {
            int right_idx = idx + (1LL << d);
            if (right_idx < n && k >= bit[right_idx]) {
                idx = right_idx;
                k -= bit[idx];
            }
        }
        return idx + 1;
    }
};

void solve() {
    int n,q; cin >> n>>q;
    vector<int> L(n);
    map<int,vector<int>> d;
    for(int i=0;i<n;i++){
        cin>>L[i];
    }
    for(int i=n-1;i>=0;i--){
        d[L[i]].push_back(i);
    }

    vector<int> k(n,0);
    for(auto &it:d){
        k[it.second[len(it.second)-1]] = 1;
        it.second.pop_back();
    }
    BIT seg = BIT(k);
    vector<tuple<int,int,int>> L2;
    for(int i=0;i<q;i++){
        int a,b;
        cin>>a>>b;
        L2.push_back(make_tuple(--a,--b,i));
    }
    sort(L2.begin(),L2.end());
    vector<int> fans(q,0);
    int k3 = 0;
    for(auto tup:L2){
        int st = get<0>(tup);
        int ed = get<1>(tup);
        int pos = get<2>(tup);
        for(int l=k3;l<st;l++){
            if(len(d[L[l]])){
                seg.update(d[L[l]][len(d[L[l]])-1],1);
                d[L[l]].pop_back();
            }
        }
        k3 = st;
        fans[pos] = seg.query(ed+1)-seg.query(st);
    }
    for(auto it:fans){
        cout<<it<<endl;
    }
}

int32_t main() {
    fastio
    int t=1;
    while (t--) solve();
    return 0;
}